File: analyzers/javascript/src/patternDetector.js
class PatternDetector {
    detectPatterns(code, astData) {
        const patterns = [];
        
        // Algorithm patterns - structure-based detection
        if (this.detectSortingPattern(code, astData)) patterns.push('sorting');
        if (this.detectBinarySearchPattern(code, astData)) patterns.push('binary_search');
        if (this.detectDivideAndConquerPattern(astData)) patterns.push('divide_and_conquer');
        if (this.detectExponentialPattern(code, astData)) patterns.push('exponential');
        
        // Specific sorting algorithms
        if (this.detectQuickSortPattern(astData)) patterns.push('quicksort');
        if (this.detectMergeSortPattern(astData)) patterns.push('mergesort');
        if (this.detectBubbleSortPattern(astData)) patterns.push('bubble_sort');
        
        // Recursion patterns
        if (astData.hasRecursion) {
            patterns.push('recursion');
            if (this.detectTailRecursion(code, astData)) patterns.push('tail_recursion');
            if (this.detectMemoization(code, astData)) patterns.push('memoization');
        }
        
        // Array patterns
        if (astData.arrayMethods.length > 0) {
            patterns.push('array_operations');
            if (this.detectMethodChaining(astData)) patterns.push('method_chaining');
        }
        
        // Async patterns
        if (astData.hasAsync) patterns.push('async_await');
        if (astData.asyncPatterns.includes('promises')) patterns.push('promises');
        
        // DOM patterns
        if (this.detectDOMManipulation(code)) {
            patterns.push('dom_manipulation');
            if (this.detectDOMInLoop(code, astData)) patterns.push('dom_in_loop');
        }
        
        // Optimization patterns
        if (this.detectDynamicProgramming(code, astData)) patterns.push('dynamic_programming');
        
        return patterns;
    }

    detectSortingPattern(code, astData) {
        // Structure-based detection
        if (astData.arrayMethods.includes('sort')) return true;
        
        // Nested loops with swap operations
        if (astData.maxLoopDepth >= 2) {
            const hasComparison = astData.binaryOperations.some(op => 
                ['<', '>', '<=', '>='].includes(op.operator) && op.context === 'in_loop'
            );
            
            // Check for array access patterns
            const hasArrayAccess = code.includes('[') && code.includes(']');
            const hasIndexAccess = /\[\s*\w+\s*\]/.test(code) && 
                                  /\[\s*\w+\s*[+-]\s*\d+\s*\]/.test(code);
            
            if (hasComparison && (hasArrayAccess || hasIndexAccess)) return true;
        }
        
        return astData.divideAndConquerSignals.includes('sorting_pattern');
    }

    detectBinarySearchPattern(code, astData) {
        // Structure-based detection
        const indicators = [
            astData.divideAndConquerSignals.includes('binary_search_pattern'),
            astData.divideAndConquerSignals.includes('division_by_2'),
            astData.loopPatterns.some(p => p.complexity === 'logarithmic')
        ];
        
        // Check for mid calculation pattern
        const hasMidCalc = /\w+[\s]*=[\s]*.*[\/>>].*2/.test(code) ||
                          /Math\.floor\s*\(.*[\/].*2\s*\)/.test(code);
        
        // Check for range updates
        const hasRangeUpdate = /\w+[\s]*=[\s]*\w+[\s]*[+-][\s]*1/.test(code);
        
        return indicators.filter(Boolean).length >= 2 || 
               (hasMidCalc && hasRangeUpdate);
    }

    detectDivideAndConquerPattern(astData) {
        const signals = astData.divideAndConquerSignals;
        
        return signals.includes('divide_and_conquer_recursion') ||
               signals.includes('range_parameters') ||
               (astData.hasRecursion && signals.includes('division_by_2'));
    }

    detectQuickSortPattern(astData) {
        // QuickSort has specific patterns: pivot selection, partitioning
        if (!astData.hasRecursion) return false;
        
        const hasPartitioning = astData.functionCalls.some(call => 
            /partition/i.test(call)
        );
        
        const hasPivot = astData.variableUpdates && 
                         Object.keys(astData.variableUpdates).some(v => /pivot/i.test(v));
        
        // Structure: recursive with multiple calls and range parameters
        const hasRangeParams = astData.divideAndConquerSignals.includes('range_parameters');
        const hasMultipleRecursion = astData.recursionPatterns.some(p => p.hasMultipleCalls);
        
        return (hasPartitioning || hasPivot) || 
               (hasRangeParams && hasMultipleRecursion && astData.maxLoopDepth >= 1);
    }

    detectMergeSortPattern(astData) {
        // MergeSort pattern: divide in half, recursive calls, merging
        if (!astData.hasRecursion) return false;
        
        const hasMerge = astData.functionCalls.some(call => 
            /merge/i.test(call)
        );
        
        const hasDivision = astData.recursionPatterns.some(p => p.hasDivision);
        const hasMultipleCalls = astData.recursionPatterns.some(p => p.hasMultipleCalls);
        
        return hasMerge || (hasDivision && hasMultipleCalls);
    }

    detectBubbleSortPattern(astData) {
        // Bubble sort: nested loops with adjacent element comparison
        if (astData.maxLoopDepth < 2) return false;
        
        // Check for swap pattern in nested loops
        const hasNestedLoops = astData.maxLoopDepth >= 2;
        const hasComparison = astData.binaryOperations.some(op => 
            ['<', '>', '<=', '>='].includes(op.operator) && op.context === 'in_loop'
        );
        
        return hasNestedLoops && hasComparison;
    }

    detectTailRecursion(code, astData) {
        // Tail recursion: return statement directly calls the function
        if (!astData.hasRecursion) return false;
        
        // Check if recursive call is in tail position
        const recursiveFuncs = astData.recursiveFunctions;
        for (const func of recursiveFuncs) {
            const pattern = new RegExp(`return\\s+${func}\\s*\\(`);  // Match return followed by function name and opening parenthesis
            if (pattern.test(code)) return true;
        }
        
        return false;
    }

    detectMemoization(code, astData) {
        // Look for caching patterns
        const cacheIndicators = [
            /memo|cache|dp/i.test(code),
            /Map|Set|Object\.create/.test(code),
            code.includes('[') && code.includes(']') && astData.hasRecursion
        ];
        
        // Check for object/array used as cache
        const hasCacheStructure = code.includes('{}') || code.includes('[]') || 
                                  code.includes('new Map') || code.includes('new Set');
        
        return cacheIndicators.filter(Boolean).length >= 2 || 
               (hasCacheStructure && astData.hasRecursion);
    }

    detectMethodChaining(astData) {
        // Check for consecutive array method calls
        const chainableMethods = ['map', 'filter', 'reduce', 'sort', 'forEach'];
        let chainCount = 0;
        
        for (let i = 0; i < astData.arrayMethods.length - 1; i++) {
            if (chainableMethods.includes(astData.arrayMethods[i]) &&
                chainableMethods.includes(astData.arrayMethods[i + 1])) {
                chainCount++;
            }
        }
        
        return chainCount >= 1;
    }

    detectDOMManipulation(code) {
        const domKeywords = [
            'document.', 'getElementById', 'querySelector', 
            'getElementsBy', 'createElement', 'appendChild',
            'innerHTML', 'textContent', 'setAttribute',
            'addEventListener', 'removeChild', 'replaceChild'
        ];
        
        return domKeywords.some(keyword => code.includes(keyword));
    }

    detectDOMInLoop(code, astData) {
        if (!this.detectDOMManipulation(code) || astData.maxLoopDepth === 0) {
            return false;
        }
        
        // More sophisticated pattern matching for DOM in loops
        const domPatterns = [
            /(for|while|forEach|map)[\s\S]{0,200}(document\.|getElementById|querySelector)/,
            /\.(forEach|map|filter|reduce)[\s\S]{0,100}(appendChild|innerHTML|textContent)/
        ];
        
        return domPatterns.some(pattern => pattern.test(code));
    }

    detectDynamicProgramming(code, astData) {
        // DP patterns: memoization, tabulation, optimal substructure
        const dpIndicators = [
            /(dp|memo|table|cache)/i.test(code),
            /Math\.(min|max)/.test(code) && astData.hasRecursion,
            astData.variableUpdates && Object.keys(astData.variableUpdates).some(v => 
                /dp|memo|table/i.test(v)
            ),
            // Check for 2D array pattern (common in DP)
            /\[\s*\w+\s*\]\s*\[\s*\w+\s*\]/.test(code)
        ];
        
        return dpIndicators.filter(Boolean).length >= 2;
    }
    // Add these methods to PatternDetector class

    detectExponentialPattern(code, astData) {
        // Check for classic exponential patterns
        
        // Fibonacci-like pattern
        if (astData.hasRecursion && astData.recursionPatterns.some(p => p.callCount >= 2)) {
            // Check if it's making multiple recursive calls with n-1, n-2 style arguments
            const hasFibPattern = /KATEX_INLINE_OPEN\s*\w+\s*-\s*1\s*KATEX_INLINE_CLOSE.*KATEX_INLINE_OPEN\s*\w+\s*-\s*2\s*KATEX_INLINE_CLOSE/.test(code) ||
                                /KATEX_INLINE_OPEN\s*\w+\s*-\s*2\s*KATEX_INLINE_CLOSE.*KATEX_INLINE_OPEN\s*\w+\s*-\s*1\s*KATEX_INLINE_CLOSE/.test(code);
            if (hasFibPattern) return true;
        }
        
        // Subset/powerset generation pattern
        if (astData.hasRecursion) {
            const hasSubsetPattern = /include|exclude|with|without/i.test(code) &&
                                    astData.recursionPatterns.some(p => p.callCount >= 2);
            if (hasSubsetPattern) return true;
        }
        
        // General exponential recursion: multiple calls without division
        return astData.recursionPatterns.some(p => p.callCount >= 2 && !p.hasDivision);
    }
}

module.exports = { PatternDetector };

File: analyzers/c/src/pattern_detector.c
#include "pattern_detector.h"
#include "ast_visitor.h"
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

static bool contains_pattern(const char* code, const char* pattern) {
    return strstr(code, pattern) != NULL;
}

static bool contains_pattern_case_insensitive(const char* code, const char* pattern) {
    size_t code_len = strlen(code);
    size_t pattern_len = strlen(pattern);
    
    if (pattern_len > code_len) return false;
    
    for (size_t i = 0; i <= code_len - pattern_len; i++) {
        bool match = true;
        for (size_t j = 0; j < pattern_len; j++) {
            if (tolower(code[i + j]) != tolower(pattern[j])) {
                match = false;
                break;
            }
        }
        if (match) return true;
    }
    
    return false;
}

bool detect_sorting_pattern(const char* code, AnalysisData* data) {
    // Bubble sort
    if ((contains_pattern_case_insensitive(code, "bubble") && 
         contains_pattern_case_insensitive(code, "sort")) ||
        (contains_pattern(code, "swap") && data->has_nested_loops)) {
        return true;
    }
    
    // Quick sort
    if ((contains_pattern_case_insensitive(code, "quick") && 
         contains_pattern_case_insensitive(code, "sort")) ||
        (contains_pattern(code, "pivot") && contains_pattern(code, "partition"))) {
        return true;
    }
    
    // Merge sort
    if ((contains_pattern_case_insensitive(code, "merge") && 
         contains_pattern_case_insensitive(code, "sort")) ||
        (contains_pattern(code, "merge") && contains_pattern(code, "divide"))) {
        return true;
    }
    
    // Heap sort
    if (contains_pattern_case_insensitive(code, "heap") && 
        contains_pattern_case_insensitive(code, "sort")) {
        return true;
    }
    
    // Check for qsort
    if (data->has_sorting) {
        return true;
    }
    
    return false;
}

bool detect_searching_pattern(const char* code, AnalysisData* data) {
    // Binary search
    bool has_mid = contains_pattern(code, "mid") || 
                   contains_pattern(code, "middle") ||
                   contains_pattern(code, "(low + high) / 2") ||
                   contains_pattern(code, "(left + right) / 2");
    
    bool has_bounds = (contains_pattern(code, "left") && contains_pattern(code, "right")) ||
                      (contains_pattern(code, "low") && contains_pattern(code, "high"));
    
    if ((contains_pattern_case_insensitive(code, "binary") && 
         contains_pattern_case_insensitive(code, "search")) ||
        (has_mid && has_bounds) ||
        (data->has_logarithmic_loop && has_bounds)) {
        return true;
    }
    
    // Check for bsearch
    if (data->has_searching) {
        return true;
    }
    
    // Linear search
    if (data->has_loops && contains_pattern(code, "==") && 
        contains_pattern(code, "return")) {
        return true;
    }
    
    return false;
}

bool detect_dynamic_programming(const char* code) {
    return contains_pattern(code, "dp[") ||
           contains_pattern(code, "memo[") ||
           contains_pattern(code, "cache[") ||
           contains_pattern(code, "table[") ||
           contains_pattern_case_insensitive(code, "memoization") ||
           contains_pattern_case_insensitive(code, "tabulation");
}

bool detect_graph_algorithm(const char* code) {
    return contains_pattern_case_insensitive(code, "dfs") ||
           contains_pattern_case_insensitive(code, "bfs") ||
           contains_pattern_case_insensitive(code, "dijkstra") ||
           contains_pattern_case_insensitive(code, "graph") ||
           contains_pattern_case_insensitive(code, "vertex") ||
           contains_pattern_case_insensitive(code, "edge") ||
           contains_pattern_case_insensitive(code, "adjacency");
}

StringArray* detect_patterns(const char* code, AnalysisData* data) {
    StringArray* patterns = create_string_array();
    if (!patterns) return NULL;
    
    // Detect algorithmic patterns
    if (detect_sorting_pattern(code, data)) {
        // Try to identify specific sorting algorithm
        if (contains_pattern_case_insensitive(code, "bubble")) {
            string_array_add(patterns, "bubblesort");
        } else if (contains_pattern_case_insensitive(code, "quick") ||
                   contains_pattern(code, "pivot")) {
            string_array_add(patterns, "quicksort");
        } else if (contains_pattern_case_insensitive(code, "merge")) {
            string_array_add(patterns, "mergesort");
        } else if (contains_pattern_case_insensitive(code, "heap")) {
            string_array_add(patterns, "heapsort");
        } else {
            string_array_add(patterns, "sorting");
        }
    }
    
    if (detect_searching_pattern(code, data)) {
        if (data->has_logarithmic_loop || contains_pattern(code, "mid") ||
            contains_pattern_case_insensitive(code, "binary")) {
            string_array_add(patterns, "binary_search");
        } else {
            string_array_add(patterns, "linear_search");
        }
    }
    
    if (detect_dynamic_programming(code)) {
        string_array_add(patterns, "dynamic_programming");
    }
    
    if (detect_graph_algorithm(code)) {
        string_array_add(patterns, "graph_algorithm");
    }
    
    // Add structural patterns from AST analysis
    if (data->has_recursion) {
        string_array_add(patterns, "recursion");
    }
    
    if (data->has_logarithmic_loop) {
        string_array_add(patterns, "logarithmic");
    }
    
    if (data->has_nested_loops) {
        string_array_add(patterns, "nested_loops");
    } else if (data->has_loops) {
        string_array_add(patterns, "single_loop");
    }
    
    if (data->has_dynamic_alloc) {
        string_array_add(patterns, "dynamic_allocation");
    }
    
    if (data->has_arrays) {
        string_array_add(patterns, "arrays");
    }
    
    // Add qsort if detected
    if (data->function_calls && string_array_contains(data->function_calls, "stdlib_sort")) {
        string_array_add(patterns, "stdlib_sort");
    }
    
    return patterns;
}